03-task-manager/starter/app.js

```diff
@@ -1 +1,37 @@
-console.log('Task Manager App')
+const express = require('express')
+const app = express()
+const port = process.env.PORT || 3000
+const tasks = require('./router/tasks')
+const connectDB = require('./DB/Connect')
+require('dotenv').config()
+const notFound = require('./middleware/Not Found')
+const errorHandler = require('./middleware/ErrorHandlerMiddle')
+
+
+//middleware
+// app.use(express.urlencoded({ extended: false }))
+// const bodyParser = require('body-parser')
+app.use(express.static('./public'))
+app.use(express.json())
+// app.use(bodyParser.json())
+
+//routes
+// app.get('/hello', (req, res) => {
+//     res.send('Task Manager App')
+// })
+
+app.use('/api/v1/tasks', tasks)
+// why is this middleware being used after our responses registered in app.use. ('/api/v1/tasks',tasks)?
+app.use(notFound)
+app.use(errorHandler)
```

Comment:
Hi @mdmontes 👋 . This is a GREAT question!

The name `middleware` is slightly confusing. In reality, the functions we pass to `app.use` (or app.get/post/etc) are really just _handlers_. In other words, they do something with the `request` and `response` objects, and optionally call `next` to let the next _handler_ do whatever it does.

Often, we pass those _handlers_ in between receiving a request initially and responding to it, hence why we refer to those as _middleware_.

But.... sometimes we pass more generic _handlers_ *after* we would have expected one of our routes to respond to a request to do some clean up. In this case, either respond with a 'not found / 404' because the request was for a route we aren't responding to; or (and this is a special use case for express) the *final* _handler_ is a bit different and takes an `error` object as the first parameter, which is the default error _handler_.

Does that help clear up why we use the `notFound` and `errorHandler` _handlers_ after the standard routes handler instead of before it?

Reply:
Hey Taylor,

Yes! This was tremendously helpful!